pkg/cmd/pulumi/search.go

```diff
@@ -0,0 +1,158 @@
+	"github.com/pulumi/pulumi/sdk/v3/go/common/env"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+	"github.com/spf13/cobra"
+	auto_table "go.pennock.tech/tabular/auto"
+)
+
+type searchCmd struct {
+	orgName     string
+	queryParams []string
+
+	Stdout io.Writer // defaults to os.Stdout
+
+	// currentBackend is a reference to the top-level currentBackend function.
+	// This is used to override the default implementation for testing purposes.
+	currentBackend func(context.Context, *workspace.Project, display.Options) (backend.Backend, error)
+}
+
+func (cmd *searchCmd) Run(ctx context.Context, args []string) error {
+	interactive := cmdutil.Interactive()
+
+	if cmd.Stdout == nil {
+		cmd.Stdout = os.Stdout
+	}
+
+	if cmd.currentBackend == nil {
+		cmd.currentBackend = currentBackend
+	}
+	currentBackend := cmd.currentBackend // shadow the top-level function
+
+	opts := backend.QueryOptions{}
+	opts.Display = display.Options{
+		Color:         cmdutil.GetGlobalColorization(),
+		IsInteractive: interactive,
+		Type:          display.DisplayQuery,
+	}
+	// Try to read the current project
+	project, _, err := readProject()
+	if err != nil && !errors.Is(err, workspace.ErrProjectNotFound) {
+		return err
+	}
+
+	backend, err := currentBackend(ctx, project, opts.Display)
+	if err != nil {
+		return err
+	}
+	cloudBackend, isCloud := backend.(httpstate.Backend)
+	if !isCloud {
+		return errors.New("Pulumi AI search is only supported for the Pulumi Cloud")
+	}
+	userName, orgs, err := cloudBackend.CurrentUser()
+	if err != nil {
+		return err
+	}
+	var filterName string
+	if cmd.orgName == "" {
+		filterName = userName
+	} else {
+		filterName = cmd.orgName
+	}
+	if !sliceContains(orgs, cmd.orgName) && cmd.orgName != "" {
+		return fmt.Errorf("user %s is not a member of org %s", userName, cmd.orgName)
+	}
+
+	parsedQueryParams := apitype.ParseQueryParams(&cmd.queryParams)
+	res, err := cloudBackend.Search(ctx, filterName, parsedQueryParams)
+	if err != nil {
+		return err
+	}
+	err = cmd.RenderTable(res.Resources)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "table rendering error: %s\n", err)
+	}
+	return nil
+}
+
+func newSearchCmd() *cobra.Command {
+	var scmd searchCmd
+	cmd := &cobra.Command{
+		Use:   "search",
+		Short: "Search for resources in Pulumi Cloud",
+		Long:  "Search for resources in Pulumi Cloud.",
+		Args:  cmdutil.NoArgs,
+		Run: cmdutil.RunFunc(func(cmd *cobra.Command, args []string) error {
+			ctx := commandContext()
+			return scmd.Run(ctx, args)
+		},
+		),
+	}
+
+	// TODO: Remove this branch once we release this feature fully.
+	if env.Dev.Value() {
+		cmd.AddCommand(newAISearchCmd())
+	}
+
+	cmd.PersistentFlags().StringVar(
+		&scmd.orgName, "org", "",
+		"Name of the organization to search. Defaults to the current user's organization.",
+	)
+	cmd.PersistentFlags().StringArrayVarP(
+		&scmd.queryParams, "query", "q", []string{},
```

Comment:
The default value can be nil. nil is a valid empty slice.

```suggestion
		&scmd.queryParams, "query", "q", nil /* default */,
```

ref: https://github.com/uber-go/guide/blob/master/style.md#nil-is-a-valid-slice